### ISSUE TITLE
Date behavior for fields that have DateOnly 

### ISSUE THREAD

**Issue by peterbjorkmarker** [microsoft/PowerPlatform-DataverseServiceClient#143]
Hi,

I have some odd behavior when using date fields with the "Date Only" behavior when using WebAPI:

             entity.Attributes["in_date"] = new DateTime(2029, 6, 1);
            var response = serviceClient.ExecuteOrganizationRequest(new CreateRequest() { Target = entity }, useWebAPI:false);

The value in the PowerApps UI is 2029,6,1.

            entity.Attributes["in_date"] = new DateTime(2029, 6, 1);
            var response = serviceClient.ExecuteOrganizationRequest(new CreateRequest() { Target = entity }, useWebAPI:true);

The value in the PowerApps UI is 2029,5,31. 

Thx!

Peter Björkmarker

**Comment by MattB-msft** [comment]
@peterbjorkmarker , 
 Interesting,  What is your user TimeZone? and your UI timezone?
we are trying to setup a repro of your specific case.

**Comment by peterbjorkmarker** [comment]
The code ran as an app user using a client secret. Timezone UTC I believe. The computer which ran the code locally is on CET (Sweden).

Thanks for the great work, I am using the client extensively with Azure functions and quite happy with it all in all. I.e. it "just works" apart from this potential bug 🙂






________________________________

**Comment by MattB-msft** [comment]
We retested this on the current clients and its working as expected.. 

With regard to the field,  "in_date" can you tell us how you have it configured? 
the type should be DateOnly, but what Behavior have you selected for it? 

Can you confirm the version of the ServiceClient your using?  we did fix a prior issue here. 

thanks !

**Comment by suodbons** [comment]
I faced the same issue. I have "Date Only" Behavior. My Timezone is UTC+3. Also, if i create something like new DateTime(2029, 6, 1, 3, 0, 0), where 3 is hours - all will be fine, in the crm it will be 2029,6,1. 

Microsoft Dynamics 365
Server version: 9.2.21071.00134
Client version: 1.4.2919-2107.1

Microsoft.PowerPlatform.Dataverse.Client.Dynamics
Version: 0.4.19

**Comment by MattB-msft** [comment]
@suodbons 
Your Field Type is DateOnly, and your behavior is DateOnly?
![image](https://user-images.githubusercontent.com/10568244/127687540-986ffcf6-4a2a-4506-86d0-932f9def6c83.png)

and what issue have you experianced?

**Comment by suodbons** [comment]
Yes, all is Date Only.
![image](https://user-images.githubusercontent.com/43095239/127689009-8d493176-6901-4070-8384-f3325067c26c.png)

I am creating some entities from files. Let's say, specific entity1 has the following reviewdate attribute:
![image](https://user-images.githubusercontent.com/43095239/127689264-f12fbde0-bb27-4047-9f02-edb487e4fcc8.png)
After organizationServise.Create(entity1) and further retrieve back I see the folowing:
![image](https://user-images.githubusercontent.com/43095239/127690033-beec416e-2b27-48b3-9efc-f941ba9943b1.png)

**Comment by suodbons** [comment]
As temp fix now I use this:

entity[attribute.Key] = TimeZoneInfo.ConvertTimeFromUtc((DateTime)attribute.Value, TimeZoneInfo.Local)

**Comment by MattB-msft** [comment]
@peterbjorkmarker  thanks for the info...  that helps!

**Comment by TclasenITVT** [comment]
It would be very helpful if the datetime kind would be properly specified instead of returning 'undefined'.
Setting it to utc for date only , date only and to local for datetime fields with timezone information.

**Comment by TclasenITVT** [comment]
In Version 5.10 this still persists. Setting a Date Only Date Only field to new Datetime(2021,1,1).ToUniversalTime() sets it to 31.12.2021 in Dataverse. You need to add at least at one hour to the datetime to ensure it is saved as the correct datetime (2021-01-01). 
My system setting is UTC+1. I think that is the factor that messes this up. I also checked the usersettings of the application user inside Dataverse, that I am using. There the timezone is set to utc.
I tested the same code on a virtual machine with utc system setting there it works as expected. Please make sure that in a future version a dataverseServiceClient Instance is not dependent on the systems datetime settings or give us a way to inject a certain timezone.

**Comment by MattB-msft** [comment]
@TclasenITVT can you set "UseWebAPI = False" on instance of the client and try again to see if you're getting proper behavior? 
we have been battling with this in the JSON serializer for a bit now.

**Comment by TclasenITVT** [comment]
Setting it directly on the client seems to fix it but using the executeOrganisationRequest(..., useWebApi:false) without the setting on the client does not.

**Comment by MattB-msft** [comment]
Ok Thanks @TclasenITVT

**Comment by MattB-msft** [comment]
As an update, 
We have been discussing this with the WebAPI team and working though the various combinations with them.  This challenge is affecting a few different clients and has flown under the over all radar for a while.   We should be getting more focus on it soon.

**Comment by peterbjorkmarker** [comment]
Hi, any news on this one?

Thanks!
Peter

**Comment by MattB-msft** [comment]
@peterbjorkmarker as we were working on entity key's we pulled this subject back up and dug back through it.. 
in retrospec, we believe the issue with the date only field being shifted is the relationship between your clock being +X and date only (no time) prevents the system from doing the conversion correctly on the client.  Without the timestamp aspect, the local client was not able to interpret the offset and move the date to the correct date ( -1 day in your case ) for the offset to work. 

We added logic to add a timestamp to the date only as part of the conversion to UTC prior to sending it in date only format in this drop. 
This MAY resolve it for you.. 

We are still testing on this.
